src/github/stage.ts

```diff
@@ -0,0 +1,97 @@
+import {context} from '@actions/github'
+import {intervalToDuration} from 'date-fns'
+import {bold} from '../slack/mrkdwn'
+import {getContextBlock} from './context'
+import {createMessage, emojiFromStatus} from './message'
+import {
+  CompletedJobStep,
+  GitHubClient,
+  isCompletedJobStep,
+  isSuccessful,
+  JobStatus,
+  Message,
+  Text
+} from './types'
+
+interface Options {
+  github: GitHubClient
+  status: string
+  now: Date
+}
+
+/**
+ * Return a progressed stage message, posted via threaded reply.
+ */
+export async function getStageMessage({
+  github,
+  status,
+  now
+}: Options): Promise<Message> {
+  const text = getText(status)
+
+  const duration = await computeDuration(github, now)
+  const contextBlock = getContextBlock(duration)
+
+  return {
+    ...createMessage(text, contextBlock),
+    reply_broadcast: !isSuccessful(status)
+  }
+}
+
+function getText(status: string): Text {
+  const verb = verbFromStatus(status)
+  const predicate = context.job
+
+  const mrkdwn = [emojiFromStatus(status), verb, bold(predicate)].join(' ')
+
+  return {
+    plain: `${verb} ${predicate}`,
+    mrkdwn
+  }
+}
+
+/**
+ * Return past tense verb for the specified job `status`.
+ */
+function verbFromStatus(status: string): string {
+  switch (status) {
+    case JobStatus.Success:
+      return 'Finished'
+    case JobStatus.Failure:
+      return 'Failed'
+    case JobStatus.Cancelled:
+      return 'Cancelled'
+    default:
+      throw new Error(`Unexpected status ${status}`)
+  }
+}
+
+async function computeDuration(
+  github: GitHubClient,
+  now: Date
+): Promise<Duration | undefined> {
+  const {data} = await github.rest.actions.listJobsForWorkflowRun({
+    ...context.repo,
+    run_id: context.runId
+  })
+
+  const slackRegex = /[^A-Za-z]slack[^A-Za-z]/i
+  const lastCompletedSlackStep = data.jobs
+    .flatMap<CompletedJobStep>(({steps}) => {
+      if (!steps) {
+        return []
+      }
+
+      return steps
+        .filter(isCompletedJobStep)
+        .filter(({name}) => slackRegex.test(` ${name} `))
+    })
+    .pop()
```

Comment:
This won't work well with _parallel_ jobs. Something to probably revisit in https://github.com/namoscato/action-slack-deploy-pipeline/issues/12.

Reply:
1. So this grab the last completed slack step, regardless of what job it is? 
2. Do you think taking in the job name as an optional parameter would solve for this?

If one is yes, I'm actually a bit surprised this works. I guess the assumption here is the jobs array returned is "ordered". I wonder however, if I have a job in the middle of my workflow but it runs last, would that break this?

Reply:
> So this grab the last completed slack step, regardless of what job it is?

Correct.

In a sequential workflow (in which jobs depend on each other), this seems to work fine; the `listJobsForWorkflowRun` response does not return jobs that have not yet started.

> Do you think taking in the job name as an optional parameter would solve for this?

Do you mean an optional input specifying the _previous_ / dependent job name? We inherently know which job we are currently in, but we actually need to know which _previous_ job / step to compute duration.

Reply:
Having given this some more thought, I _am_ leaning towards scoping "stage" duration to its contextual job. We'll still want to support multiple Slack steps per job, but I now see less value in computing this across jobs, especially given the threaded message displays the job's name.

Reply:
Updated in c6eb5f9bec65dd26fe211cd9766f63d888d0f246.

It still favors a `lastCompletedSlackStep` but scoped within the current job, which still allows for scenarios in which multiple slack notifications can be sent within the same job, i.e.

```yaml
steps:
- name: Post to Slack
- name: Deploy to staging
- name: Post to Slack
- name: Deploy to prod
- name: Post to Slack
```